File: 380-insert-delete-getrandom-o1/insert-delete-getrandom-o1.py
class RandomizedSet:

    def __init__(self):
        self.numbers = []
        self.valToIndex = {}

    def insert(self, val: int) -> bool:
        if val in self.valToIndex:
            return False
        self.valToIndex[val] = len(self.numbers)
        self.numbers.append(val)
        return True

    def remove(self, val: int) -> bool:
        n = len(self.numbers)
        if val not in self.valToIndex:
            return False
        index_to_remove = self.valToIndex[val]
        last_element = self.numbers[-1]
        if index_to_remove == n - 1:
            self.removeLastElement(val)
            return True
        self.numbers[index_to_remove] = last_element
        self.valToIndex[last_element] = index_to_remove
        self.removeLastElement(val)
        return True

    def removeLastElement(self, lastVal: int):
        self.numbers.pop()
        del self.valToIndex[lastVal]

    def getRandom(self) -> int:
        return self.numbers[randint(0, len(self.numbers) - 1)]


# Your RandomizedSet object will be instantiated and called as such:
# obj = RandomizedSet()
# param_1 = obj.insert(val)
# param_2 = obj.remove(val)
# param_3 = obj.getRandom()

File: 380-insert-delete-getrandom-o1/insert-delete-getrandom-o1.java
class RandomizedSet {

    private Map<Integer, Integer> valToIndexMap;
    private List<Integer> numbers;

    private Random random;

     public RandomizedSet() {
        this.valToIndexMap = new HashMap<>();
        this.numbers = new ArrayList<>();
        this.random = new Random();
    }

    public boolean insert(int val) {
         if (this.valToIndexMap.containsKey(val)){
             return false;
         }
         this.valToIndexMap.put(val, numbers.size());
         numbers.add(val);
         return true;
    }

    public boolean remove(int val) {
         if(!this.valToIndexMap.containsKey(val)){
             return false;
         }
         int lastElement = this.numbers.get(this.numbers.size() - 1);
         int indexToRemove = this.valToIndexMap.get(val);
         this.valToIndexMap.put(lastElement, indexToRemove);
         this.numbers.set(indexToRemove, lastElement);
         this.numbers.remove(this.numbers.size() - 1);
         this.valToIndexMap.remove(val);
         return true;
    }

    public int getRandom() {
         int randomIndex = this.random.nextInt(this.numbers.size());
         return this.numbers.get(randomIndex);
    }
}

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * RandomizedSet obj = new RandomizedSet();
 * boolean param_1 = obj.insert(val);
 * boolean param_2 = obj.remove(val);
 * int param_3 = obj.getRandom();
 */